### ISSUE TITLE
Feature request: conditionally merge duplicated members

### ISSUE THREAD

**Issue by wmjordan** [gluck/il-repack#215]
Is it possible to remove some members during the repack?
For instance, I'd like to remove some members marked with a specific attribute, or given full names.

**Comment by gluck** [comment]
No this feature isn't available, not sure it makes sense for the tool, though there's already options for flagging types as intern, what's your use case for removing these members ?

**Comment by wmjordan** [comment]
I am attempting to merge two assemblies compiled from different projects.
One from C#, the other one from a dynamically generated assembly, or IL.
The projects contain dummy types to keep them from cross-referencing each other.
If the remove members function is ready, it is possible to merge members from those two assemblies into a single type.

**Comment by deniszykov** [comment]
I think it is easier for you to write custom merge tool with Mono.Cecil

**Comment by gluck** [comment]
You mean you are using allowdup/union to merge these types with same name into one ?
And getting rejects because they have members with same name/signature ?

**Comment by gluck** [comment]
(actually ILRepack ignores duplicated members, so if your "true" members are in the same assembly, providing it first in the command line should work)

**Comment by wmjordan** [comment]
@gluck Thanks for replying.
You are right, I am using /allowdup to merge types with same name into one.
Rather than relying on the order of assemblies, is it possible to specify
`[System.Runtime.CompilerServices.MethodImpl(
System.Runtime.CompilerServices.MethodImplOptions.ForwardRef)]`
on the method to tell ILRepack that the annotated method is the "fake" one?

**Comment by wmjordan** [comment]
> so if your "true" members are in the same assembly, providing it first in the command line should work

When I was trying to use this approach, I met with a problem about the `AssemblyVersion`.

My primary assembly named A needed to be merged with another assembly named B.
In A, there were several methods which should be overridden by their counterparts in B.

If I specify B first, followed by A, then A and B got merged together, however, the AssemblyVersion was the one in B, no longer the on in the primary assembly, A, even if I have specified `/attr:A.dll` in the command line.
If I specify A first, then members in B could not be merged into A.

**Comment by wmjordan** [comment]
Well, I solved the above problem by specifying `/ver:$(AssemblyVersion)` in the _csproj_ file which called ILRepack.

**Comment by gluck** [comment]
lots of topics there ! Here's my take:
- `/attr` [should ignore](https://github.com/gluck/il-repack/blob/0f26607100844f611fe3fb2f738424b3e4cea21d/ILRepack/Steps/AttributesRepackStep.cs#L65) what you have in any assembly, so the custom attributes (incl. version) are likely correct, but the IL assembly version hasn't been corrected from the attributes (bug I'd say, as they should stay in sync)
- `ForwardRef` support sounds interesting, so that if there's a duplicate member, the one w/o the attribute will take precedence 👍

**Comment by wmjordan** [comment]
The problem of `ForwardRef` is that it can be applied to methods only.

Maybe it is possible to take advantage of the `[ObfuscationAttribute]` to instruct ILRepack to merge duplicated members, for that attribute can be applied to all members, and if possible, we can even centralize the merging of duplicate members in the _Property\AssemblyInfo.cs_ file.
After repacking, the relative `[Obfuscation]` attributes can be stripped out.

**Comment by gluck** [comment]
I'd rather support an attribute by name only (like mono-linker `[Preserve]` [attribute](https://github.com/gluck/il-repack/blob/linker/ILRepack/Steps/Linker/ApplyPreserveAttribute.cs#L51)) rather than abuse the obfuscation one (unrelated IMO).
E.g. we could support `[ForwardRef]` by name, if the official one works for you, use it, if it doesn't, create your own.

Or we could support another by-name-attributes, like `RepackDrop` that would ignore the member during repack.

**Comment by wmjordan** [comment]
OK. I am looking forward to it.

**Comment by wmjordan** [comment]
How are you doing @gluck ? Any progress on this?

**Comment by gluck** [comment]
Hello, sorry if I let you think otherwise, but I have no plan to implement this myself.
I'd welcome a PR along the lines of what we discussed though if you need that feature !

Let me know.

**Comment by gluck** [comment]
Hello, sorry if I let you think otherwise, but I have no plan to implement this myself.
I'd welcome a PR along the lines of what we discussed though if you need that feature !

Let me know.

**Comment by wmjordan** [comment]
I found the "RepackDrop" annotation support was quite easy to implement. I took several minutes and I got it up and running.
Since there was only one file which needed to be changed. I uploaded the source code here.
Please take it if it is appropriated.
[RepackImporter.zip](https://github.com/gluck/il-repack/files/1770729/RepackImporter.zip)

**Comment by wmjordan** [comment]
Simply add a method and use it in the `Import` method, things are done.

````C#
private bool CanInclude<TMember> (TMember member)
	where TMember : ICustomAttributeProvider, IMemberDefinition {
	// skip members marked with a custom attribute named "RepackDrop"
	var marked = member.HasCustomAttributes
		&& member.CustomAttributes.FirstOrDefault(attr => attr.AttributeType.Name == "RepackDropAttribute") != null;
	if (marked) {
		_logger.Log("Repack dropped " + member.FullName);
	}
	return marked == false;
}
````

**Comment by wmjordan** [comment]
Usage: 
1. Manually define an attribute named `RepackDropAttribute`, and use it on the member we need to exclude from repacking.
1. Run ILRepack as usual.

**Comment by wmjordan** [comment]
The PR added option support for the _RepackDrop_ attribute name (by default, `RepackDropAttribute`)

**Comment by wmjordan** [comment]
Implemented via #221
